### DISCUSSION TITLE
Rename the Azure module and reimagine the Azure support

### DISCUSSION THREAD

**Discussion by nagyesta** [testcontainers/testcontainers-java#9607] [Ideas]
Hi,
first of all, thank you for working on the great project, it saved me a lot of work over the years! 
In fact, I am considering to open some PRs to add some more Azure modules which I think would be beneficial. But first, I wanted to ask whether that is something you would be okay with... Let me to explain the idea...

As I have seen, the Azure support is somewhat limited in the Java repository but has more useful modules in other languages. I think it could be great to have feature parity with the others. To be able to do that, we would need to follow the steps below:

### 1. Rename: Azure -> CosmosDB
I have noticed, that the module naming is different here, maybe it would be a good idea to consider renaming the Azure module to [CosmosDB](https://github.com/testcontainers/testcontainers-java/blob/main/modules/azure/src/main/java/org/testcontainers/containers/CosmosDBEmulatorContainer.java) which is the only supported container in it. This would help the Java support to be in line with the [.Net](https://github.com/testcontainers/testcontainers-dotnet/tree/develop/src/Testcontainers.CosmosDb) and [Python](https://github.com/testcontainers/testcontainers-python/tree/main/modules/cosmosdb) projects.

### 2. Additional modules

Then we could add more modules similarly to the other languages as shown below:

| .Net | Go | Node | Python |
| ----- | ---| ------ | -------- |
| [Azurite](https://github.com/testcontainers/testcontainers-dotnet/tree/develop/src/Testcontainers.Azurite) | [Azurite](https://github.com/testcontainers/testcontainers-go/tree/main/modules/azurite) | [Azurite](https://github.com/testcontainers/testcontainers-node/tree/main/packages/modules/azurite) | [Azurite](https://github.com/testcontainers/testcontainers-python/tree/main/modules/azurite) |
| [Service Bus](https://github.com/testcontainers/testcontainers-dotnet/tree/develop/src/Testcontainers.ServiceBus) | | | |

---

### The benefits

With this, Java developers could benefit from all 3 modules as well (Azurite, CosmosDB, Service Bus). Plus they can use [my third party module to emulate Azure Key Vault with Lowkey Vault](https://github.com/nagyesta/lowkey-vault/tree/main/lowkey-vault-testcontainers) if anyone would like.

> [!NOTE] 
> I am not asking you to make that module part of the project, because I don't qualify as a reputable source, violating the first rule of module submission right away:
> > Should be a Docker Hub official image, or published by a reputable source (ideally the company or organisation that officially supports the technology)

### The problem

Of course, this would be a breaking change as the Azure module will be renamed.

### The questions

- What do you think? 
- Would you be okay if I opened PRs implementing any part of this plan? 
- Do you have an alternative solution in mind?

Thank you in advance for your kind consideration!

**Comment by eddumelendez** [comment]
Hi, thanks for bringing this topic. Azure module should contain implementations for Azure products such as Cosmosdb, Azurite, EventHubs and ServiceBus emulators.

Regardless of multiple modules or one module the user would have access to all implementations provided by Testcontainers.

Adding `AzuriteContainer` should be straightforward. See https://github.com/eddumelendez/testcontainers-samples/blob/589ac4d32c013cc66b7f26b83bff5ea81bb629ec/spring-cloud-azure-eventhubs/src/test/java/com/example/springcloudazureeventhubs/SpringCloudAzureEventHubsApplicationTests.java#L45-L48. PR is welcome

EventHubs depends on Azurite. See https://github.com/eddumelendez/testcontainers-samples/blob/main/spring-cloud-azure-eventhubs/src/test/java/com/example/springcloudazureeventhubs/SpringCloudAzureEventHubsApplicationTests.java#L45-L62. The expected behavior would be to allow declare `EventHubsContainer` and under the hood will start `AzuriteContainer`. Also, declare and external azurite instance if needed. Due to API limitations I would go with only first approach. PR is welcome

ServiceBus depends on Azure SQL Edge. See https://github.com/Azure/azure-service-bus-emulator-installer/blob/main/Docker-Compose-Template/docker-compose-default.yml. Similar to what is described for EventHubs. However, I haven't explore Azure SQL Edge, looking at [Testcontainers for .NET](https://github.com/testcontainers/testcontainers-dotnet/blob/develop/src/Testcontainers.ServiceBus/ServiceBusBuilder.cs#L114-L124), the MSSQL module should help. AFAIK, the Azure ServiceBus doesn't support the emulator url yet, so we should wait for it. See https://github.com/Azure/azure-service-bus-emulator-installer/issues/25

**Reply by nagyesta** [reply]
Hi @eddumelendez ,
thank you, I will look into it and send PRs when ready!

**Reply by nagyesta** [reply]
Hi @eddumelendez ,
FYI: I finally managed to create a PR for the Azurite Container, please find it here: #9661

**Reply by eddumelendez** [reply]
I have added an [example](https://github.com/eddumelendez/testcontainers-samples/blob/main/spring-cloud-azure-servicebus/src/test/java/com/example/springcloudazureservicebus/SpringCloudAzureServiceBusApplicationTests.java) using ServiceBus emulator. azure sdk has been released supporting emulator endpoint

**Reply by nagyesta** [reply]
Thanks, I will take a look when I have time.

**Reply by nagyesta** [reply]
Hi, FYI: I have just created a PR for the Service Bus Emulator: https://github.com/testcontainers/testcontainers-java/pull/9795
